### DISCUSSION TITLE
Would there be any way to make this `merge` function work?

### DISCUSSION THREAD

**Discussion by scarf005** [gvergnaud/ts-pattern#279] [Q&A] [answered]
```ts
import { match, P } from 'ts-pattern'

const merge = <A>(xs: A[], ys: A[]): A[] =>
    match([xs, ys])
        .with([P._, []], ([xs, _]) => xs)
        .with([[], P._], ([_, ys]) => ys)
        .with([[P._, ...P.array()], [P._, ...P.array()]], ([x, xs_, y, ys_]) =>
            x <= y ? [x, ...merge(xs_, ys_)] : [y, ...merge(xs_, ys_)])
        .exhaustive()

```

**Playground Link:** [Provided](https://www.typescriptlang.org/play/?jsx=0#code/JYWwDg9gTgLgBAbziAhjAxgCwDRwApwC+cAZlBCHAOQwDOAtGGjAKZQB2VAUF+hO7Xgg2AcxZwAvHAA8AQQB8ACgAetAFxxZAbQC6uAJ7rNugJQbtOyfK5xbyNFkVbVB2jpM27XgHQB3YDCYTnjeAPq4unpwTi5woe5WcKoeXj7+gU66uCHxuE7hcIYJEvKFtCmptn4BQVpaObjeTSEoUFAo+oomUfVhjc3ere2d3VExuKoF+q7xJlaelV7KMlL6cAD8cM793sJQYiq0U0fdcBpa03BNu6Ish8eh3RWV3izKmCgAroLAAG53HiAA)
      
- it's the `merge` function from mergesort.
- it fails both in type level and in runtime because it matches into `null`.

**Comment by gvergnaud** [comment, marked as answer]
It's a known limitation that TS-Pattern doesn't handle matching on type parameters very well. That's because TypeScript inference gets stuck on unknown type parameters, and there is unfortunately nothing we can do about it in userland.

You can work around this by using an overload for your merge function: [Playground](https://www.typescriptlang.org/play/?jsx=0#code/JYWwDg9gTgLgBAbziAhjAxgCwDRwApwC+cAZlBCHAOQwDOAtGGjAKZQB2VAUFyQK7t0MYBHbI2AcxYAeAIIA+ABQAPWgC44sgNoBdXAE91m3QEoN2nbwFCRYkJJYqj7PiABGbXQeeuPUUxou7p46iFxwEXBQLDB8HMhoWIpaqt46JuGRWQB0AO7AMJjJeNkA+ri6enDJqXCl6XAAvPJwqhlZOfmFyV74ZVXJ5XCGDc3DtO0dEXkFRVpaJUPZyyUoUFAo+oomVQtluMvZq+ub2zoD88oHy6r1FfrX2Yb1o-KZU1nKcNKNw3AA-HAUo97FApE4hs8dnANFoHnBDqDwbdvKUdpMptkWMpMCg+LRhAA3RwZQhAA)

This also works, but is more brittle because it relies on implementation details of the way inference works (if you don't use parameters, their types dont get evaluated): [Playground](https://www.typescriptlang.org/play/?jsx=0&ssl=1&ssc=1&pln=2&pc=1#code/JYWwDg9gTgLgBAbziAhjAxgCwDRwApwC+cAZlBCHAOQwDOAtGGjAKZQB2VAUF+hO7Xgg2AcxZwAvHAA8AQQB8ACgAetAFxxZAbQC6uAJ7rNugJQbtOyfK5xbyNFkVbVB2jpM27XgHRQWMAFcOABV9MBY5XSUPLx8Ad2AYTCc8bwB9XF09OCcXODT3KzhVGNjbbwSkp11cVILcJwy4Q0KJeWbaUrKKxOStLTrcb2HUlCgoFH1FE2yB9KGR7zGJqZns6aKETzKvPgF4ZwXvVQLJYtptnds9wTgtfSPDU6lDS6u-QI5zmRe4AH47sojsIoGIVLQmk8ZnANPdgaIWODIRCZm87IQurFvCxlJgUAFBMAAG6IjxAA)
